### PR TITLE
Add reparameterization support to `OneHotCategorical`

### PR DESCRIPTION
**Pull request by lqf96**
Add reparameterization support to the `OneHotCategorical` distribution. Samples are reparameterized based on the straight-through gradient estimator, which is proposed in the paper [Estimating or Propagating Gradients Through Stochastic Neurons for Conditional Computation](https://arxiv.org/abs/1308.3432).

### CODE REVIEW

**Comment by zou3519** [comment]
@neerajprad, @fritzo, could you take a look at this please?

**Comment by fritzo** [comment]
Hmm the idea is sound, but I believe this may break a lot of software if we make `OneHotCategorical` reparametrized by default. I think it would be safe to either (1) add a `has_rsample` kwarg to the constructor and default to False, or (2) create a separate distribution say `OneHotCagetoricalStraightThrough` that inherits from `OneHotCategorical`. I believe @karalets has done something [similar](http://docs.pyro.ai/en/stable/distributions.html#relaxedonehotcategoricalstraightthrough) in Pyro . Does anyone else have an interface opinion?

**Comment by lqf96** [comment]
@fritzo Personally I think that depends on if there are more than one way to do reparameterization for `OneHotCateogorical`. If there is only one strightforward way to do it (i.e. straight-through gradient) then it probably makes sense to simply add an `enable_rsample` parameter. If there are more than one common way to do it, we should go with the `OneHotCagetoricalStraightThrough` API instead.

File: torch/distributions/one_hot_categorical.py [hunk 1]
```diff
@@ -74,11 +79,20 @@ def param_shape(self):
         return self._categorical.param_shape
 
     def sample(self, sample_shape=torch.Size()):
+        # work around the issue of `Tensor.expand` not respecting `torch.no_grad`
+        return self.rsample(sample_shape).detach()
```
Comment (fritzo) [inline torch/distributions/one_hot_categorical.py:83]: I think we should avoid the extra overhead of `.add(probs).sub(probs).detach()`. How about this refactoring:
```py
def sample(self, sample_shape=torch.Size()):
    # ...unchanged by this PR...

def rsample(self, sample_shape=torch.Size()):
    samples = self.sample(sample_shape)
    probs = self._categorical.probs  # note this is cached via @lazy_property
    return samples + (probs - probs.detach())
```

**Review by fritzo — CHANGES_REQUESTED** [review]
Since this is approximate and since there are multiple ways to approximate the gradients [1], I believe it would be best to create a subclass
```py
class OneHotCategoricalStraightThrough(OneHotCategorical):
    has_rsample = True
    def rsample(self, sample_shape=torch.size()):
        samples = self.sample(sample_shape)
        probs = self._categorical.probs  # note this is cached via @lazy_property
        return samples + (probs - probs.detach())
```
@martinjankowiak @eb8680 @karalets does this seem reasonable to you?

[1] Bengio et al (2013) https://arxiv.org/abs/1308.3432

<details>

> Another estimator of the expected gradient through stochastic neurons was proposed by Hinton (2012) in his lecture 15b. The idea is simply to back-propagate through the hard threshold function (1 if the argument is positive, 0 otherwise) as if it had been the identity function. It is clearly a biased estimator, but when considering a single layer of neurons, it has the right sign (this is not guaranteed anymore when back-propagating through more hidden layers). We call it the straight-through (ST) estimator. A possible variant investigated here multiplies the gradient on hi by the derivative of the sigmoid. Better results were actually obtained without multiplying by the derivative of the sigmoid.

</details>

**Comment by martinjankowiak** [comment]
yes, i think subclassing is definitely the way to go here 👍

File: torch/distributions/one_hot_categorical.py (hunk 1 above)
Comment (lqf96) [reply torch/distributions/one_hot_categorical.py:83]: This looks great and I'll refactor the code over the weekend.

**Comment by lqf96** [comment]
@fritzo I applied your suggestions and I think this is ready for review again. There are some spurious test failures and hopefully they're not related to this PR.

**Review by fritzo — APPROVED** [review]
LGTM. Test failures appear unrelated. Thanks for your patience.
